lib/cn/cn_tree.c

```diff
@@ -1965,20 +1961,263 @@ static merr_t
+            }
+        }
+
+    } else {
+        err = cn_subspill_commit(ss);
+    }
 
-    err = cn_subspill_commit(ss);
     if (err)
         return err;
 
-    if (ss->ss_added) {
-        w->cw_output_nodev[0] = ss->ss_node;
-        w->cw_checkpoint(w);
-    }
+    w->cw_output_nodev[0] = ss->ss_node;
+    w->cw_checkpoint(w);
 
     return 0;
 }
 
+/* Tree must be locked when this is called */
+struct cn_tree_node *
+cn_kvset_can_zspill(struct kvset *ks, struct route_map *map)
+{
+    unsigned char maxkbuf[HSE_KVS_KEY_LEN_MAX];
+    const void *maxkey, *maxptkey, *minkey;
+    uint16_t maxklen, maxptklen, minklen;
+
+    struct route_node *minnode;
+
+    kvset_minkey(ks, &minkey, &minklen);
+    kvset_maxkey(ks, &maxkey, &maxklen);
+    kvset_max_ptkey(ks, &maxptkey, &maxptklen);
+
+    if (maxptkey && keycmp_prefix(maxptkey, maxptklen, maxkey, maxklen) >= 0) {
+        maxklen = maxptklen;
+
+        memcpy(maxkbuf, maxptkey, maxptklen);
+        memset(&maxkbuf[maxptklen], 0xff, sizeof(maxkbuf) - maxptklen);
+        maxkey = maxkbuf;
+    }
+
+    minnode = route_map_lookup(map, minkey, minklen);
+    assert(minnode);
+
+    if (route_node_keycmp(minnode, maxkey, maxklen) >= 0)
+        return route_node_tnode(minnode);
+
+    return NULL;
+}
+
+/* TODO: This function shares a fair bit of code with cn_comp_spill() and should be refactored.
+ */
+static merr_t
+cn_zspill(struct cn_compaction_work *w, bool *zspilled)
+{
+    struct kvset_list_entry *first, *le = w->cw_mark;
+    struct cn_tree *tree = w->cw_tree;
+    struct cn_tree_node *n, *znode;;
+    struct subspill *ss_arr;
+    struct route_node *rtn = NULL;
+    atomic_uint *spillingp = NULL;
+    atomic_uint *zspillingp = NULL;
+    void *lock;
+    uint i;
+    merr_t err = 0;
+
+    *zspilled = false;
+    znode = NULL;
+
+    rmlock_rlock(&tree->ct_lock, &lock);
+
+    /* Confirm that the operation can still be a zspill.
+     */
+    for (i = 0, first = le; i < w->cw_kvset_cnt; ++i) {
+        first = le;
+
+        assert(kvset_get_workid(le->le_kvset) != 0);
+        n = cn_kvset_can_zspill(le->le_kvset, tree->ct_route_map);
+        if (!n)
+            break;
+
+        if (!znode)
+            znode = n;
+        else if (n->tn_nodeid != znode->tn_nodeid)
+            break;
+
+        le = list_prev_entry(le, le_link);
+    }
+
+    if (i < w->cw_kvset_cnt) {
+        rmlock_runlock(lock);
+        return 0; /* Input kvsets no longer map to a single node. Cannot zspill, bail out */
+    }
+
+    /* Cannot zspill if znode is splitting or joining.
+     */
+    mutex_lock(&tree->ct_ss_lock);
+    if (znode->tn_ss_splitting || znode->tn_ss_joining) {
+        mutex_unlock(&tree->ct_ss_lock);
+        rmlock_runlock(lock);
+        return 0;
+    }
+
+    /* Since znode is neither splitting nor joining, the operation can continue as a zspill. Mark
+     * the target child node of the zspill as busy so it may be protected from future splits and
+     * joins.
+     */
```

Comment:
I would replace the sentence starting with "Mark.." with "Increment znode's spill ref to prevent split/join until after the zspill has completed."